### ISSUE TITLE
Any support for using texture arrays?

### ISSUE THREAD

**Issue by xaxxon** [ocornut/imgui#571]
I just transitioned my program from using plain 2d textures to using 2d texture arrays.   Now I don't think I can use imGui::ImageButton and the like, since it has no support for layer.  

Assuming that's true, do you know of any sample code or examples on how to add this type of support?

I can think of plenty of workarounds, of course, just wondering if there's something I'm missing.

Thank you.

**Comment by ocornut** [comment]
> I just transitioned my program from using plain 2d textures to using 2d texture arrays. Now I don't think I can use imGui::ImageButton and the like, since it has no support for layer.

I don't understand what any of this means, nor what texture arrays are (is that OpenGL terminology?) nor why it would affect anything wih imgui, please clarify. 

All Image()/ImageButton() do are passing a `void*` that you provide to a renderer that you provide.

**Comment by xaxxon** [comment]
Sorry about that.

A texture array (aka a 3d texture) is a set of equal sized 2d ("normal") textures addressed as UVW where "w" is the index into the texture array.  

The entire index is given a single opengl texture name (GLuint) and in your frag shader you say something like:

```
uniform sampler2DArray textures;
in vec3 texture_coordinates;

out vec4 color;

void main(){
	 color = texture(textures, texture_coordinates);
}

```

Note how `texture_coordinates` is a vec3, not a vec2.   

I am not good at opengl so anything I said above might not be exactly correct.

I guess instead of just putting in the simple GLint for the texture into the ImTextureID I would put both the GL texture name and the Z coordinate.   I forgot that was a void*.

**Comment by ocornut** [comment]
Thanks for the explanation. It would make sense to carry more data into the `void*` than just the GLint then. Depending how many layers you have, it might have sense to store an array of { gluint, z } in your high-level texture format (to avoid requiring the user to store that structure in a persistent manner), and use a pointer to this structure as your ImTextureId.

If you want non-integer Z for interpolation between two layers (which may or not be possible? I'm guessing) then you don't have a finite number of values so user would need to store it. If you are running 64-bit only then `void*` is guaranteed to fit a GLuint and a float to it would be a trick/possibility to encode both into the `void*`.

The Z part can be passed as a uniform to your shader and the shader would create a vec3() out of the vec2() from imgui vertex buffer + the z passed as uniform, so you don't need to modify the underlying data structure for the imgui vertex and it would be cheaper overall. You can technically override the layout for ImDrawVert with `IMGUI_OVERRIDE_DRAWVERT_STRUCT_LAYOUT` and add an extra `float w` there but it would be overkill.

**Comment by xaxxon** [comment]
It doesn't seem like sampler2darray with w=0 works with a "normal" texture.   So that means it would need a different program in some situations ..

I'll try to figure out more about opengl though before I mess around too much with imgui.   There's probably some other solution that I can't even imagine because of my limited ogl knowledge.

**Comment by ocornut** [comment]
Yeah you might need to switch program depending on if you want to sample from a sampler2d or a sampler2darray. Said, store w=-1 in your texture id for non array types of texture for example?

Or there could be a way to map an array slice into a regular texture, I don't know either! Good luck :)

**Comment by xaxxon** [comment]
it actually turned out to be far simpler than I thought.   When I cobbled together my "integration layer" with bits form the SDL fixed function opengl example and the glfw opengl3 example (I'm using SDL and opengl3+), I didn't know what I was doing and just mixed and matched parts until it worked.   Going back and looking at the code, it makes a lot more sense now and I didn't realize imgui only uses that one static texture.   I just changed that one texture to be a 1-element texture array, passed in a pointer to a struct that has both the texture name (GLuint) and depth to the io.Fonts->TexID and then since the rendering is a separate draw call for each texture, just set the depth as a uniform.

My shaders ended up like this;

    const GLchar *vertex_shader =
            "#version 330\n"
                    "uniform mat4 ProjMtx;\n"
**                    "uniform uint Layer;\n"
**                    "in vec2 Position;\n"
                    "in vec2 UV;\n"
                    "in vec4 Color;\n"
                    "out vec3 Frag_UV;\n"
                    "out vec4 Frag_Color;\n"
                    "void main()\n"
                    "{\n"
**                    "	Frag_UV = vec3(UV, Layer);\n"
**                    "	Frag_Color = Color;\n"
                    "	gl_Position = ProjMtx * vec4(Position.xy,0,1);\n"
                    "}\n";

    const GLchar* fragment_shader =
            "#version 330\n"
**                    "uniform sampler2DArray Texture;\n"
****                    "in vec3 Frag_UV;\n"
**                    "in vec4 Frag_Color;\n"
                    "out vec4 Out_Color;\n"
                    "void main()\n"
                    "{\n"
                    "	Out_Color = Frag_Color * texture( Texture, Frag_UV);\n"
                    "}\n";

Maybe someone else will see this thread and it will help them out :)
